api/src/npm/tarball.rs

```diff
@@ -101,10 +101,23 @@ pub fn create_npm_tarball<'a>(
       transpiled_files.insert(path.to_owned(), json.source.as_bytes().to_vec());
     } else if let Some(js) = module.js() {
       match js.media_type {
-        deno_ast::MediaType::JavaScript
-        | deno_ast::MediaType::Mjs
-        | deno_ast::MediaType::Dts
-        | deno_ast::MediaType::Dmts => {
+        // We need to transpile js files too, to rewrite import sources
+        // from `npm:*` to bare specifiers, for example.
+        deno_ast::MediaType::JavaScript | deno_ast::MediaType::Mjs => {
+          let source = sources
+            .get_parsed_source(module.specifier())
+            .expect("parsed source should be here");
+          let source_url = Url::options()
+            .base_url(Some(registry_url))
+            .parse(&format!("./@{scope}/{package}/{version}{path}",))
+            .unwrap();
+          let transpiled = transpile_to_js(source, source_url)
+            .with_context(|| format!("failed to transpile {}", path))?;
```

Comment:
This is kind of annoying. Means source maps also become necessary for JS.

I'll do a follow up on this to make this use text changes, like we do in CLI, rather than going via the AST. That way we can avoid having to re-emit the JS.